// @framer/dropboxdesign.desktop-kit/code/Subcomponents/MessagePreview.tsx
import * as React from "react"
import { addPropertyControls, ControlType, MotionStyle, Frame } from "framer"
import { useUnsplashAPI } from "../Utils/fetch"

MessagePreview.defaultProps = {
  keyword: "Restaurant",
  title: "Dinner at Maido - Top 13 Restaurant in the world by Food Quest",
  url: "foodquest.com"
}

export function MessagePreview(props) {
  const { width, height, keyword, title, url } = props

  const { imageUrl, isImageLoading } = useUnsplashAPI({
    size: "600x400",
    search: keyword
  })

  return (
    <Frame style={backgroundStyle(isImageLoading ? "" : imageUrl)}>
      <div style={footerStyle}>
        <div style={titleStyle}>{title}</div>
        <div style={urlStyle}>{url}</div>
      </div>
    </Frame>
  )
}

addPropertyControls(MessagePreview, {
  keyword: {
    type: ControlType.String,
    title: "Keyword",
    placeholder: "Pick a keyword"
  },
  title: {
    type: ControlType.String,
    title: "Title"
  },
  url: {
    type: ControlType.String,
    title: "URL"
  }
})

const backgroundStyle = (imageUrl): MotionStyle => {
  return {
    position: "relative",
    height: "100%",
    width: "100%",
    backgroundImage: `url(${imageUrl})`,
    backgroundPosition: "center",
    backgroundSize: "cover"
  }
}

const footerStyle: React.CSSProperties = {
  position: "absolute",
  bottom: 0,
  left: 0,
  right: 0,
  padding: "8px 14px 10px",
  fontFamily: "-apple-system, system-ui, BlinkMacSystemFont, Segoe UI",
  background: "#E5E6E9"
}

const titleStyle: React.CSSProperties = {
  fontSize: 13,
  color: "#000"
}

const urlStyle: React.CSSProperties = {
  fontSize: 10,
  paddingTop: 4,
  color: "#8F8F8F"
}
