// @framer/edoardo.dot-grid/code/Dots.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";

interface Props {
  dotColor: string;
  bgColor: string;
  dotSize: number;
  dotSpace: number;
}

export class Dots extends React.Component<Props> {
  static defaultProps = {
    width: 198,
    height: 198,
    dotColor: "#fff",
    bgColor: "#000",
    dotSize: 2,
    dotSpace: 20
  };

  static propertyControls: PropertyControls<Props> = {
    dotSize: { type: ControlType.Number, min: 1, max: 10, title: "Size" },
    dotSpace: { type: ControlType.Number, min: 1, max: 100, title: "Spacing" },
    dotColor: { type: ControlType.Color, title: "Color" },
    bgColor: { type: ControlType.Color, title: "Background" }
  };

  render() {
    let { dotColor, bgColor, dotSize, dotSpace } = this.props;

    return (
      <div
        style={{
          height: "100%",
          background: `linear-gradient(90deg, ${bgColor} ${dotSpace -
            dotSize}px, transparent 1%) top left / ${dotSpace}px ${dotSpace}px, linear-gradient(${bgColor} ${dotSpace -
            dotSize}px, transparent 1%) top left / ${dotSpace}px ${dotSpace}px ${dotColor}`
        }}
      />
    );
  }
}
