// @framer/dropboxdesign.desktop-kit/code/Subcomponents/WebView.tsx
import * as React from "react"
import { useState } from "react"
import { ControlType, addPropertyControls, Frame } from "framer"
import { Helpers_DefaultInstructions as DefaultInstructions } from "../canvas"
import { Spinner } from "./Spinner"

/* -------------------------------------------------------------------------- */
/*                                   WebView                                  */
/* -------------------------------------------------------------------------- */

WebView.defaultProps = {
    width: 1440,
    height: 900,
    url: "www.example.com",
}

export function WebView(props) {
    const { width, height, url, onMouseDown } = props
    const [loading, setLoading] = useState(true)

    const handleOnLoad = React.useCallback(() => {
        setLoading(false)
    }, [])

    const src = `http://${url.replace(/(^\w+:|^)\/\//, "")}/`
    const frameBorder = 0
    const MemoizedIframe = React.memo(props =>
        React.createElement("iframe", props)
    )

    return url !== "" ? (
        <Frame style={webViewStyle(width, height)}>
            {loading ? <Spinner /> : null}
            <MemoizedIframe
                src={src} // Sanitize the url if it contains 'http://' and such
                onLoad={handleOnLoad}
                style={iframeStyle}
                frameBorder={frameBorder}
                sandbox={"allow-scripts"}
                name={"IFRAME_" + src}
            />
        </Frame>
    ) : (
        <DefaultInstructions
            Message="Enter a valid URL for this WebView. Certain major websites (like Google.com) may not load due to content restrictions. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options)"
            size="100%"
            Arrow=""
        />
    )
}
const webViewStyle = (width, height) => {
    return {
        width,
        height,
        backgroundColor: "transparent",
    }
}
const iframeStyle = { width: "100%", height: "100%" }

addPropertyControls(WebView, {
    url: {
        type: ControlType.String,
        title: "URL",
        placeholder: "www.example.com",
    },
})
